azure_costs_exporter/scrape.py

```diff
@@ -53,17 +53,27 @@ def convert_json_df(data):
 
     return df
 
-def extract_metrics_from_df(df, counter):
+def extract_metrics_from_df(df, metric_name, enrollment, registry):
     """
     Fill a counter object with the data from the data frame
     """
     groups = df.groupby(base_columns).sum()
-    for name, value in groups.iterrows():
-        meta = dict(zip(base_columns, name))
-        counter.record(value.ExtendedCost, **meta)
 
+    class AzureCollector(object):
+        def collect(self):
+            c = CounterMetricFamily(metric_name,
+                                    "Costs billed to Azure Enterprise Agreement {}".format(enrollment),
+                                    labels=base_columns)
 
-def query_metrics(enrollment, token, metric_name, month=None):
+            for name, value in groups.iterrows():
+                c.add_metric(name, value.ExtendedCost)
+
+            yield c
+
+    registry.register(AzureCollector())
```

Comment:
So:
* Make the collector a top-level class. The configuration values could be arguments to the initializer
* In its `collect()` method, gather new data, manipulate it, and return it in terms of metrics
* Use `query_metrics` and `extract_metrics_from_df` as helpers of your collector (members), not vice versa
* Register the metrics exactly _once_, and do this externally